### ISSUE TITLE
clean compilation with -Wextra

### ISSUE THREAD

**Issue by gnzlbg** [jemalloc/jemalloc#1196]
Currently jemalloc doesn't compile free of warnings with `-Wextra` enabled in GCC and clang (https://github.com/jemalloc/jemalloc/issues/1019#issuecomment-328819406)

-Wextra has helped diagnose bugs in the past:

* https://github.com/jemalloc/jemalloc/issues/1019#issuecomment-328819406

It might be worth it to fix the warnings being reported here and raise the level from -Wall to -Wextra for those two compilers.

**Comment by sldevel** [comment]
> * locally suppresses some -Wextra diagnostics:
>
>  * `-Wmissing-field-initializer` is buggy in older Clang and GCC versions,
>   where it does not understanding that, in C, `= {0}` is a common C idiom
>    to initialize a struct to zero

Err... did you actually *test* the resulting code ?
Rationale: if the compiler warns about an idiom it does not understand, then risks are that this idiom doesn't lead to any code produced by the said compiler. In that case, it would mean that the structure is never zeroed when compiled with such "old" (please, define "old") compilers...

**Comment by gnzlbg** [comment]
> Err... did you actually test the resulting code ?

Yes, I did  - previous versions of this PR actually "fixed" this manually, but that was extremely verbose, so we decided to revert those changes and just suppress the warning for this case. In C `= {0}` initializes all fields of a struct to zero. I managed to track the GCC bug about this (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36750), and the only thing buggy AFAICT was the warning (which is/was part of -Wextra and not -Wall probably because of these issues). I did not looked for a similar bug in clang though, so what you mention is possible, but given how common the idiom is, I'd say it is extremely unlikely.

> please, define "old"

The bug was first reported for GCC 4.1 in 2008, and the first fix (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36750#c3) landed in 2011, but fixed the warning for C only (the GCC version in which the fix landed was not reported there, but from looking at the dates of the GCC releases probably the first version with the fix was GCC 4.6.1). The C++ version of the warning was still broken until somewhere between GCC 4.9.2 and GCC 5.1.1 (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=36750#c13). 

---

FWIW I think that unconditionally suppressing the warning might be a too big of a hammer, and what we could do is suppress the warning only for those compilers in which it is broken. PRs welcome I guess.
